### PR TITLE
ramips: add support for I-O DATA WN-AC733GR3

### PR DESCRIPTION
**Pull request by musashino205**
I-O DATA WN-AC733GR3 is a 2.4/5 GHz band 11ac router, based on
MediaTek MT7620A.

Specification

- SoC		: MediaTek MT7620A
- RAM		: DDR2 64 MiB
- Flash		: SPI-NOR 8 MiB
- WLAN		: 2.4/5 GHz
  - 2.4 GHz : MT7620A (SoC), 2T2R
  - 5 GHz   : MT7610E, 1T1R
- Ethernet	: 10/100/1000 Mbps (RTL8367RB)
- LED/key	: 4x/4x (2x buttons, 1x slide-switch)
- UART		: through-hole on PCB
  - J1: Vcc, RX, GND, TX from LED side
  - 57600n8

Flash instruction using factory image:

1. Boot WN-AC733GR3 normaly
2. Access to "http://192.168.0.1/" and open firmware update page
("ファームウェア")
3. Select the OpenWrt factory image and click update ("更新") button
to perform firmware update
4. Wait ~150 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

### CODE REVIEW

File: target/linux/ramips/mt7620/config-4.14 [hunk 1]
```diff
@@ -200,6 +200,8 @@ CONFIG_RALINK_WDT=y
 # CONFIG_RCU_STALL_COMMON is not set
 CONFIG_REGMAP=y
 CONFIG_REGMAP_MMIO=y
+CONFIG_REGULATOR=y
```
Comment (diizzyy) [inline target/linux/ramips/mt7620/config-4.14]: This should be broken out as a separate PR

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 2]
```diff
@@ -196,7 +196,8 @@ ht-tm02)
 	ucidef_set_led_netdev "eth" "Ethernet" "$boardname:green:lan" "eth0"
 	set_wifi_led "$boardname:blue:wlan"
 	;;
-iodata,wn-ac1167gr)
+iodata,wn-ac1167gr|\
+iodata,wn-ac733gr3)
```
Comment (diizzyy) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: Please use alphabetical order

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 3]
```diff
@@ -324,7 +324,8 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "5:wan" "6@eth0"
 		;;
-	iodata,wn-ac1167gr)
+	iodata,wn-ac1167gr|\
+	iodata,wn-ac733gr3)
```
Comment (diizzyy) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Please use alphabetical order

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 4]
```diff
@@ -546,7 +547,8 @@ ramips_setup_macs()
 		[ -n "$lan_mac" ] || lan_mac=$(cat /sys/class/net/eth0/address)
 		wan_mac=$(macaddr_add "$lan_mac" 1)
 		;;
-	iodata,wn-ac1167gr)
+	iodata,wn-ac1167gr|\
+	iodata,wn-ac733gr3)
```
Comment (diizzyy) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Please use alphabetical order

**Comment by diizzyy** [comment]
Not sure if you should numerical order instead of alphabetical order

File: target/linux/ramips/dts/WN-AC733GR3.dts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+		auto {
+			label = "auto";
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+			linux,input-type = <EV_SW>;
+		};
+	};
+
+	regulator {
```
Comment (chunkeey) [inline target/linux/ramips/dts/WN-AC733GR3.dts]: Is this really a power line regulator for the rtl8367rb? Or is this regulator here just so it can be used to control the rtl8367rb's reset line?

Because if this is just the reset line, then we are in a bit of a pickle. On one hand, you could do away with the regulator here (so you don't need to add the symbol in the kernel config) and go straight for a gpio-hog instead. 

Or on the other hand do a bit of preparation. Because the upstream linux kernel driver will do this differently, since it has a "reset-gpios" property that fits perfectly: https://www.kernel.org/doc/Documentation/devicetree/bindings/net/dsa/realtek-smi.txt

File: target/linux/ramips/dts/WN-AC733GR3.dts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	};
+};
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 {
+	wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
```
Comment (chunkeey) [inline target/linux/ramips/dts/WN-AC733GR3.dts]: Please add the missing compatible. (I guess it's missing in the template you work from as well. But let's work on just one issue at a time).

The options here are either "pciVVVV,PPPP" (where VVVV is the 16-bit vendor id and PPPP is the 16-bit product it) or the compatible listed in the binding document:

https://www.kernel.org/doc/Documentation/devicetree/bindings/net/wireless/mediatek%2Cmt76.txt

File: target/linux/ramips/dts/WN-AC733GR3.dts (hunk 5 above)
Comment (musashino205) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: I think that it is probably a reset line. (I cannot be confident because I cannot find the document for RTL8367RB...)
OK, I'll try to use gpio-hog.

Implementation in Linux Kernel, I have been interested for a while. However, I do not know how to configure the interrupt controller correctly.

File: target/linux/ramips/dts/WN-AC733GR3.dts (hunk 6 above)
Comment (musashino205) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: I see, I'll add compatible.

File: target/linux/ramips/mt7620/config-4.14 (hunk 1 above)
Comment (musashino205) [reply target/linux/ramips/mt7620/config-4.14]: OK

**Comment by musashino205** [comment]
updated:
  - use gpio-hog instead of regulator
    - drop kernel config for regulator
  - add compatible into wifi node

File: target/linux/ramips/dts/WN-AC733GR3.dts (hunk 5 above)
Comment (chunkeey) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: >I think that it is probably a reset line. (I cannot be confident because I cannot find the document for RTL8367RB...)

Well, If you want to measure it to confirm whenever this is the reset line or not: According to my RTL8367RB.pdf (found by google, same chip is used in the WNDAP620/660 and a few other) the nRESET signal is supposed to be on Pin 91 of the LQFP-128 package. This would be the 6th pin from the top left corner of the chip. if said chip is oriented in such a way that the Pin 1 marker is in the bottom/lower-left corner (i.e.: the text and the realtek crab has the "proper" orientation).

Comment (musashino205) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: Thank you for the information, I'll check it.

Comment (chunkeey) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: > Thank you for the information, I'll check it.

Forgot to add the datasheets comment on the 91 pin: "When low active will reset the RTL8367RB. ".

As for power pins: there are several analog and digital voltages ranging from 3.3, 2.5 and 1.0 volts.

Comment (musashino205) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: Thank you, I'll keep in mind.

**Comment by chunkeey** [comment]
> Not sure if you should numerical order instead of alphabetical order

Well, this is what busybox "case" statement does when it's matching the argument with the case lists. 

From what I can tell, this happens in the "casematch" function which checks the argument with every single case list in the order there were placed in the script file (by line).
https://github.com/mirror/busybox/blob/master/shell/ash.c#L9259

now casematch itself is just a delegating the match to
- patmatch
https://github.com/mirror/busybox/blob/master/shell/ash.c#L9259
- pmatch:
https://github.com/mirror/busybox/blob/master/shell/ash.c#L7910
- fnmatch
https://github.com/mirror/busybox/blob/master/shell/ash.c#L6211
- musl' fnmatch
http://git.musl-libc.org/cgit/musl/tree/src/regex/fnmatch.c#n306

which finally seems to do the real comparison by starting from the beginning
and then checks byte by byte. 

So if we wanted to optimize this for something like "best avg. response for everyone worldwide". we would need to run a popularity contest and sort the board entries in the 01_leds / 02_network by popularity instead.  (And schedule reruns from time to time :microscope:)

But this will likely consume more power than it will be worth it and there will likely be a endless struggle for :1st_place_medal:. So I think that because we are dealing with a mix of numbers, letters and symbols in a delicious but confusing soup, the lexicographical order (that would follow the workflow of the fnmatch function) is the best choice :tm: .

File: target/linux/ramips/dts/WN-AC733GR3.dts (hunk 5 above)
Comment (musashino205) [reply target/linux/ramips/dts/WN-AC733GR3.dts]: result (GPIO 0):
- set LOW  -> RTL8367RB RESET: 0 V
- set HIGH -> RTL8367RB RESET: 3.3V

**Comment by chunkeey** [comment]
Merged. Thanks.
